stor/dx.py

```diff
@@ -385,129 +1067,161 @@ def exists(self):
+        except (stor_exceptions.NotFoundError, ValueError):
             pass
-        # otherwise we could be a directory, so try to grab first
-        # file/subfolder
-        if self._is_folder():
-            try:
-                self.list(limit=1)
-                return True
-            except NotFoundError:
-                return False
-        return False
+        # otherwise we could be a directory, so try to listdir folder
+        # note list doesn't error on non-existent folder and cannot be used here
+        try:
+            self.listdir()
+            return True
+        except stor_exceptions.NotFoundError:
+            return False
 
+    def getsize(self):
+        if not self.resource:
+            return self.stat()['dataUsage']*1e9
+        else:
+            return self.stat()['size']
+
+    @_propagate_dx_exceptions()
     def stat(self):
-        if not self.__stat:
+        """Performs a stat on the path.
 
-            if self.canonical_resource:
-                self.__stat = dxpy.DXFile(dxid=self.canonical_resource,
-                                          project=self.canonical_project).describe()
-            else:
-                self.__stat = dxpy.DXProject(dxid=self.canonical_project).describe()
-        return self.__stat
+        Raises:
+            DuplicateError: If project or resource is not unique
+            NotFoundError: When the project or resource cannot be found
+            ValueError: If path is folder path
+        """
+        if not self.resource:
+            return dxpy.DXProject(dxid=self.canonical_project).describe()
+        return dxpy.DXFile(dxid=self.canonical_resource,
+                           project=self.canonical_project).describe()
 
 
 class DXVirtualPath(DXPath):
-    """Class Handler for DXPath of form 'dx://project-{ID}:/a/b/c' or 'dx://a/b/c'"""
+    """Class Handler for DXPath of form 'dx://project-{dxID}:/a/b/c' or 'dx://a:/b/c'"""
 
     @property
     def virtual_project(self):
+        """Returns the virtual name of the project associated with the DXVirtualPath"""
         if utils.is_valid_dxid(self.project, 'project'):
-            return dxpy.DXProject(dxid=self.project).name
+            with _propagate_dx_exceptions():
+                return dxpy.DXProject(dxid=self.project).name
         return self.project
 
     @property
     def virtual_resource(self):
+        """Returns the human readable path of the resource associated with the DXVirtualPath"""
         return self.resource
 
     @property
     def virtual_path(self):
+        """Returns the DXVirtualPath instance"""
         return self
 
     @cached_property
     def canonical_project(self):
-        """Returns the unique project that matches the name that user has view access to.
-        If no match is found, returns None
+        """Returns the ID of the unique project that matches the name that user has at least
+        view access to. If no match is found, returns None
```

Comment:
```suggestion
        only resolves project user has access to.
```

You say `NotFoundError` is raised when project doesn't exist on DNAnexus but doc above says no match found then returns None

Reply:
Changed. The docstring was outdated. It definitely raises a `NotFoundError`